regression/cbmc-incr-smt2/bitvector-arithmetic-operators/overflow_behaviour.c

```diff
@@ -0,0 +1,11 @@
+#include <limits.h>
+#include <stdlib.h>
+
+int main() {
+    int a = INT_MAX;
```

Comment:
I see that we have tests for `int` and `unsigned int`. These both have the same width. We should test with varying widths of integer, rather than just a single width.

Reply:
I'm not sure how to deliver this within the current implementation.

I have played with a variety of signed and unsigned integers (`int8_t`, `uint16_t`, etc) and it always seems to hit a invariant failure related to typecasting:

```
File: /Users/fotiskoutoulakis/Devel/cbmc/src/solvers/smt2_incremental/convert_expr_to_smt.cpp:63 function: convert_expr_to_smt
Condition: false
Reason: Reached unimplemented Generation of SMT formula for type cast expression: typecast
  * type: signedbv
      * width: 32
      * #c_type: signed_int
  0: symbol
      * type: unsignedbv
          * width: 16
          * #typedef: uint16_t
          * #c_type: unsigned_short_int
      * identifier: main::1::c!0@1#1
      * expression: symbol
          * type: unsignedbv
              * width: 16
              * #typedef: uint16_t
              * #c_type: unsigned_short_int
          * #source_location: 
            * file: ../regression/cbmc-incr-smt2/bitvector-arithmetic-operators/unsigned_behaviour.c
            * line: 15
            * function: main
            * working_directory: /Users/fotiskoutoulakis/Devel/cbmc/build_newops
          * identifier: main::1::c
      * #SSA_symbol: 1
      * L0: 0
      * L1: 1
      * L2: 1
      * L1_object_identifier: main::1::c!0@1
Backtrace:
0   cbmc                                0x00000001014e7ac0 _Z15print_backtraceRNSt3__113basic_ostreamIcNS_11char_traitsIcEEEE + 68
1   cbmc                                0x00000001014e8014 _Z13get_backtracev + 44
2   cbmc                                0x0000000100297b94 _Z29invariant_violated_structuredI17invariant_failedtJRKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEEENS1_9enable_ifIXsr3std10is_base_ofIS0_T_EE5valueEvE4typeES9_S9_iS9_DpOT0_ + 56
3   cbmc                                0x0000000100297b28 _ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEC1IDnEEPKc + 0
4   cbmc                                0x00000001010e7370 _ZL19convert_expr_to_smtRK14typecast_exprt + 164
5   cbmc                                0x00000001010e6530 _Z19convert_expr_to_smtRK5exprt + 148
6   cbmc                                0x00000001010f8ac4 _ZL25convert_relational_to_smtIN30smt_function_application_termt8factorytIN22smt_bit_vector_theoryt31unsigned_greater_than_or_equaltEEENS1_INS2_29signed_greater_than_or_equaltEEEE9smt_termtRK21binary_relation_exprtRKT_RKT0_ + 280
7   cbmc                                0x00000001010e90f4 _ZL25try_relational_conversionRK5exprt + 176
8   cbmc                                0x00000001010e685c _Z19convert_expr_to_smtRK5exprt + 960
9   cbmc                                0x0000000101102f7c _ZN36smt2_incremental_decision_proceduret30ensure_handle_for_expr_definedERK5exprt + 192
10  cbmc                                0x00000001011032f0 _ZN36smt2_incremental_decision_proceduret6handleERK5exprt + 132
11  cbmc                                0x0000000100880a28 _ZN22symex_target_equationt19convert_assumptionsER19decision_proceduret + 308
12  cbmc                                0x00000001008802a8 _ZN22symex_target_equationt26convert_without_assertionsER19decision_proceduret + 140
13  cbmc                                0x0000000100881690 _ZN22symex_target_equationt7convertER19decision_proceduret + 60
14  cbmc                                0x00000001002d6c70 _Z29convert_symex_target_equationR22symex_target_equationtR19decision_proceduretR16message_handlert + 116
15  cbmc                                0x00000001002d86b4 _Z24prepare_property_deciderRNSt3__13mapI8dstringt14property_infotNS_4lessIS1_EENS_9allocatorINS_4pairIKS1_S2_EEEEEER22symex_target_equationtR28goto_symex_property_decidertR19ui_message_handlert + 208
16  cbmc                                0x0000000100318d30 _ZN25multi_path_symex_checkert24prepare_property_deciderERNSt3__13mapI8dstringt14property_infotNS0_4lessIS2_EENS0_9allocatorINS0_4pairIKS2_S3_EEEEEE + 44
17  cbmc                                0x0000000100318c04 _ZN25multi_path_symex_checkertclERNSt3__13mapI8dstringt14property_infotNS0_4lessIS2_EENS0_9allocatorINS0_4pairIKS2_S3_EEEEEE + 400
18  cbmc                                0x00000001002d3a3c _ZN43all_properties_verifier_with_trace_storagetI25multi_path_symex_checkertEclEv + 44
19  cbmc                                0x00000001002b3884 _ZN19cbmc_parse_optionst4doitEv + 6088
20  cbmc                                0x000000010155db94 _ZN19parse_options_baset4mainEv + 1328
21  cbmc                                0x0000000100295b84 main + 60
22  dyld                                0x00000001064710f4 start + 520
```

This appears to be because we don't yet have support for `typecast_exprt`.

At first I thought this might be because of automatic conversion of literals used in the conditions, but even if I isolate them out and initialise a variable of type, say `int8_t` with `0`, it still comes across the same issue.

Reply:
I guess we can't use the platform independent `<stdint.h>` header yet, due to it introducing type casts. It may be possible to work-around this by using the type `long long int` and the `ll` suffix for literals. Or by using the `char` type and character constants.

For types which vary in width between platforms (such as `int`), you can make the regexes pass over all environments by matching either of the correct values for the 2 different widths we expect.